### PR TITLE
docs: initial stdlib contributor docs

### LINKED ISSUE #464: [Docs]: Add a contributor document for our stdlib patterns
We've kinda been nailing down the stdlib patterns, so I think it's time we write it down.

### PR DESCRIPTION
**Pull request by phated**
This is a very rough draft of contributor docs for our standard library.

Closes #464 

@ospencer @peblair I think this needs a lot of input from you both. Maybe we can use Suggestions to collaboratively edit this!

### CODE REVIEW

File: docs/contributor/standard_library.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Standard Library
+
+Grain aims to have a comprehensive and consistent standard library.
+
+To us, this means that someone should be able to reach for the standard library to perform most of their day-to-day work, and it will always work the way they expect in the context of Grain.
```
Comment (ospencer) [inline docs/contributor/standard_library.md:5]: Love this.

File: docs/contributor/standard_library.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+## Guidelines
+
+Here are some guidelines for making additions to the standard library while also keeping it high-quality and consistent!
+
+1. New data types should exist in their own modules. For example, the `Range` enum is the data type exported from `range.gr`. An exception to this are data types that are ubiquitous (`Option`/`Result`) or require compiler support (`List`).
```
Comment (ospencer) [inline docs/contributor/standard_library.md:11]: `List` doesn't actually get much for compiler support—I think it lives in Pervasives for the same reasons that Option and Result do. You just wanna be able to make them/annotate using them without having to import anything.
```suggestion
1. New data types should exist in their own modules. For example, the `Range` enum is the data type exported from `range.gr`. An exception to this are data types that are ubiquitous (`Option`/`Result`/`List`), which should live in `pervasives.gr`.
```

File: docs/contributor/standard_library.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+1. Prefer data constructors when possible. Only use separate constructor methods, like `make` or `init`, if needed to set an initialization value or to hide internals of your data structure.
+1. All methods that operate on a data type should exist in the same module as that data type. For example, all `Map` methods exist in the `map.gr` file.
+1. Fallible methods should almost always return an `Option` or `Result`. Usage of `fail` should be reserved for exeptional cases, such as argument validation producing an index out-of-bounds failure.
+    * `Option` should be preferred if you might or might not be able to get some value.
```
Comment (ospencer) [inline docs/contributor/standard_library.md:15]: Maybe we describe this as something that might be nullable in other languages?

File: docs/contributor/standard_library.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    * `Result` is useful if you have multiple failures, or if the consumer might need additional context around the failure.
```
Comment (ospencer) [inline docs/contributor/standard_library.md:16]: Should we add examples to these, like `List.find` should return `Option` and `Number.parseInt` should return `Result`?

**Review by ospencer — COMMENTED** [review]
I think this is great! Thank you for putting this together 😄

File: docs/contributor/standard_library.md (hunk 3 above)
Comment (phated) [reply docs/contributor/standard_library.md]: Hows that update look?

File: docs/contributor/standard_library.md (hunk 4 above)
Comment (phated) [reply docs/contributor/standard_library.md]: Hows that update look?

**Review by ospencer — APPROVED** [review]
Updates look great!! Thanks for putting this together 😊

File: docs/contributor/standard_library.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+Grain aims to have a comprehensive and consistent standard library.
+
+To us, this means that someone should be able to reach for the standard library to perform most of their day-to-day work, and it will always work the way they expect in the context of Grain.
+
+## Guidelines
+
+Here are some guidelines for making additions to the standard library while also keeping it high-quality and consistent!
+
+1. New data types should exist in their own modules. For example, the `Range` enum is the data type exported from `range.gr`. An exception to this are data types that are ubiquitous (`Option`/`Result`/`List`), which should live in `pervasives.gr`.
+1. Prefer data constructors when possible. Only use separate constructor methods, like `make` or `init`, if needed to set an initialization value or to hide internals of your data structure.
```
Comment (peblair) [inline docs/contributor/standard_library.md:12]:
```suggestion
1. Prefer data constructors when possible. Only use separate constructor functions, like `make` or `init`, if needed to set an initialization value or to hide internals of your data structure.
```
Grain doesn't have methods; we only have functions.

File: docs/contributor/standard_library.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+1. All methods that operate on a data type should exist in the same module as that data type. For example, all `Map` methods exist in the `map.gr` file.
```
Comment (peblair) [inline docs/contributor/standard_library.md:13]:
```suggestion
1. All functions that operate on a data type should exist in the same module as that data type. For example, all `Map` methods exist in the `map.gr` file.
```

File: docs/contributor/standard_library.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+1. Fallible methods should almost always return an `Option` or `Result`. Usage of `fail` should be reserved for exeptional cases, such as argument validation producing an index out-of-bounds failure.
```
Comment (peblair) [inline docs/contributor/standard_library.md:14]:
```suggestion
1. Fallible functions should almost always return an `Option` or `Result`. Usage of `fail` should be reserved for exceptional cases, such as argument validation producing an index out-of-bounds failure.
```

File: docs/contributor/standard_library.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    * `Option` should be preferred if you might or might not be able to get some value. Typically, these would be methods that could return `null` in other languages when they didn't produce a value. For example, `List.find` returns `None` when no item in the list matches the condition.
```
Comment (peblair) [inline docs/contributor/standard_library.md:15]:
```suggestion
    * `Option` should be preferred if you might or might not be able to get some value. Typically, these would be functions that could return `null` in other languages when they didn't produce a value. For example, `List.find` returns `None` when no item in the list matches the condition.
```

File: docs/contributor/standard_library.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    * `Result` is useful if you have multiple failures, or if the consumer might need additional context around the failure. Typically, these methods would throw exceptions in other languages. For example, `Number.parseInt` returns `Err(reason)` when it fails to parse a string into an integer.
```
Comment (peblair) [inline docs/contributor/standard_library.md:16]:
```suggestion
    * `Result` is useful if you have multiple failures, or if the consumer might need additional context around the failure. Typically, these functions would throw exceptions in other languages. For example, `Number.parseInt` returns `Err(reason)` when it fails to parse a string into an integer.
```

**Review by peblair — CHANGES_REQUESTED** [review]
This is great! Just a minor semantic nit and one typo.

**Comment by phated** [comment]
@peblair good catch! All suggestions applied 😄

**Pull request merged**